decentraland/unity-explorer
# fix: Terrain overallocation and LHO

## What does this PR change?

When generating the terrain, the `TerrainLocalCache` contains several dictionaries used as persisent objects until we call the `Save()` method to "flush" them.  Once saved, even if I called `Clear()` or `new Dictionary()` I couldnt retrieve the memory.

Investigating a little bit it looks like the issue is related to the management of Large Heap Objects. When an objects is over 85K bytes, it falls into this special heap. And `TerrainLocalCache` was well over it.

This apparently generated a fragmentation that was impossible to recover, leaving 700MB of unusable space. If it was recoverable, clearing the dictionaries should have returned me that space, but that was not the case.

So, this PR saves the dictionaries into smaller files. This means that we are free to use an additional 700MB.

There is a still big difference between generating the cache and reading from the cache, almost 1.5GB. It would be ideal that the cache generation consumed the same as the read (as the result is the same). We need to investigate it further.

Here are some results. This is the resultant memory after getting to GP

-------

`main`

**CACHE GENERATION 9.0GB**
<img width="1728" alt="Screen Shot 2024-09-24 at 15 56 44" src="https://github.com/user-attachments/assets/d788471a-b131-4f3e-a594-c348972695e9">

**CACHE READ 7.6GB**
<img width="1728" alt="Screen Shot 2024-09-24 at 15 55 08" src="https://github.com/user-attachments/assets/1b739480-bd1b-4363-99d1-fca9c011b837">

-------

`pr`

**CACHE GENERATION 8.2GB**
<img width="1728" alt="Screen Shot 2024-09-24 at 20 50 10" src="https://github.com/user-attachments/assets/979d0f67-855d-4920-82cf-7e0177dd02fb">

**CACHE READ 6.7GB**
<img width="1728" alt="Screen Shot 2024-09-24 at 20 52 25" src="https://github.com/user-attachments/assets/4ee1ba35-fcce-42c0-88ca-28833af50f01">


## How to test the changes?



1. Launch the explorer with no terrain cache (Look for [`Application.persistentDataPath`](https://docs.unity3d.com/ScriptReference/Application-persistentDataPath.html) for files that have the `terrain_cache` name)
2. Generate the terrain and fly around. Try to remember as much as possible
3. Now restart. Terrain should look the same

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
